### PR TITLE
Prickly Pufferfish

### PR DESCRIPTION
**Pull request by Tru-Dev**
This is our team's submission! It will be a code battle site where people code versus each other!

### CODE REVIEW

**Comment by SebastiaanZ** [comment]
Needs manual merge:
- `.DS_Store` file was added to the root of the repo

---

flake8 linting errors

```
./prickly-pufferfish/exec/test_random_question.py:3:1: F403 'from functions import *' used; unable to detect undefined names
./prickly-pufferfish/exec/test_random_question.py:14:30: F405 'add_to_zero' may be undefined, or defined from star imports: functions
./prickly-pufferfish/exec/test_random_question.py:15:30: F405 'add_to_zero' may be undefined, or defined from star imports: functions
./prickly-pufferfish/exec/test_random_question.py:16:30: F405 'add_to_zero' may be undefined, or defined from star imports: functions
./prickly-pufferfish/exec/test_random_question.py:17:30: F405 'add_to_zero' may be undefined, or defined from star imports: functions
./prickly-pufferfish/exec/test_random_question.py:20:30: F405 'is_anagram_of_palindrome' may be undefined, or defined from star imports: functions
./prickly-pufferfish/exec/test_random_question.py:21:30: F405 'is_anagram_of_palindrome' may be undefined, or defined from star imports: functions
./prickly-pufferfish/exec/test_random_question.py:22:30: F405 'is_anagram_of_palindrome' may be undefined, or defined from star imports: functions
./prickly-pufferfish/exec/test_random_question.py:23:30: F405 'is_anagram_of_palindrome' may be undefined, or defined from star imports: functions
./prickly-pufferfish/exec/test_random_question.py:24:30: F405 'is_anagram_of_palindrome' may be undefined, or defined from star imports: functions
./prickly-pufferfish/exec/test_random_question.py:27:30: F405 'coins' may be undefined, or defined from star imports: functions
./prickly-pufferfish/exec/test_random_question.py:28:30: F405 'coins' may be undefined, or defined from star imports: functions
./prickly-pufferfish/exec/test_random_question.py:29:30: F405 'coins' may be undefined, or defined from star imports: functions
./prickly-pufferfish/exec/test_random_question.py:30:30: F405 'coins' may be undefined, or defined from star imports: functions
./prickly-pufferfish/exec/test_random_question.py:31:30: F405 'coins' may be undefined, or defined from star imports: functions
./prickly-pufferfish/exec/test_random_question.py:32:43: E128 continuation line under-indented for visual indent
./prickly-pufferfish/exec/test_random_question.py:35:29: F405 'has_balanced_brackets' may be undefined, or defined from star imports: functions
./prickly-pufferfish/exec/test_random_question.py:36:29: F405 'has_balanced_brackets' may be undefined, or defined from star imports: functions
./prickly-pufferfish/exec/test_random_question.py:37:29: F405 'has_balanced_brackets' may be undefined, or defined from star imports: functions
./prickly-pufferfish/exec/test_random_question.py:38:29: F405 'has_balanced_brackets' may be undefined, or defined from star imports: functions
./prickly-pufferfish/exec/test_random_question.py:39:30: F405 'has_balanced_brackets' may be undefined, or defined from star imports: functions
./prickly-pufferfish/exec/test_random_question.py:40:30: F405 'has_balanced_brackets' may be undefined, or defined from star imports: functions
./prickly-pufferfish/exec/test_random_question.py:41:30: F405 'has_balanced_brackets' may be undefined, or defined from star imports: functions
./prickly-pufferfish/exec/test_random_question.py:42:30: F405 'has_balanced_brackets' may be undefined, or defined from star imports: functions
./prickly-pufferfish/exec/test_random_question.py:45:30: F405 'make_change' may be undefined, or defined from star imports: functions
./prickly-pufferfish/exec/test_random_question.py:46:30: F405 'make_change' may be undefined, or defined from star imports: functions
./prickly-pufferfish/exec/test_random_question.py:49:30: F405 'merge_ranges' may be undefined, or defined from star imports: functions
./prickly-pufferfish/exec/test_random_question.py:50:42: E128 continuation line under-indented for visual indent
./prickly-pufferfish/exec/test_random_question.py:51:30: F405 'merge_ranges' may be undefined, or defined from star imports: functions
./prickly-pufferfish/exec/test_random_question.py:52:42: E128 continuation line under-indented for visual indent
./prickly-pufferfish/exec/test_random_question.py:53:30: F405 'merge_ranges' may be undefined, or defined from star imports: functions
./prickly-pufferfish/exec/test_random_question.py:54:30: F405 'merge_ranges' may be undefined, or defined from star imports: functions
./prickly-pufferfish/exec/test_random_question.py:55:42: E127 continuation line over-indented for visual indent
./prickly-pufferfish/exec/test_random_question.py:56:30: F405 'merge_ranges' may be undefined, or defined from star imports: functions
./prickly-pufferfish/exec/test_random_question.py:59:30: F405 'get_max_profit' may be undefined, or defined from star imports: functions
./prickly-pufferfish/exec/test_random_question.py:60:30: F405 'get_max_profit' may be undefined, or defined from star imports: functions
./prickly-pufferfish/exec/test_random_question.py:61:30: F405 'get_max_profit' may be undefined, or defined from star imports: functions
./prickly-pufferfish/exec/test_random_question.py:62:30: F405 'get_max_profit' may be undefined, or defined from star imports: functions
./prickly-pufferfish/exec/test_random_question.py:63:30: F405 'get_max_profit' may be undefined, or defined from star imports: functions
./prickly-pufferfish/exec/test_random_question.py:64:30: F405 'get_max_profit' may be undefined, or defined from star imports: functions
./prickly-pufferfish/exec/test_random_question.py:67:30: F405 'valid_parens_perms' may be undefined, or defined from star imports: functions
./prickly-pufferfish/exec/test_random_question.py:68:30: F405 'valid_parens_perms' may be undefined, or defined from star imports: functions
./prickly-pufferfish/exec/test_random_question.py:69:30: F405 'valid_parens_perms' may be undefined, or defined from star imports: functions
./prickly-pufferfish/exec/test_random_question.py:72:30: F405 'valid_parens_perms' may be undefined, or defined from star imports: functions
./prickly-pufferfish/exec/test_random_question.py:82:30: F405 'zigzag' may be undefined, or defined from star imports: functions
./prickly-pufferfish/exec/test_random_question.py:83:30: F405 'zigzag' may be undefined, or defined from star imports: functions
./prickly-pufferfish/exec/test_random_question.py:84:30: F405 'zigzag' may be undefined, or defined from star imports: functions
./prickly-pufferfish/exec/test_random_question.py:85:30: F405 'zigzag' may be undefined, or defined from star imports: functions
./prickly-pufferfish/exec/test_random_question.py:86:30: F405 'zigzag' may be undefined, or defined from star imports: functions
./prickly-pufferfish/exec/test_random_question.py:87:30: F405 'zigzag' may be undefined, or defined from star imports: functions
./prickly-pufferfish/exec/test_random_question.py:88:30: F405 'zigzag' may be undefined, or defined from star imports: functions
./prickly-pufferfish/exec/test_random_question.py:89:30: F405 'zigzag' may be undefined, or defined from star imports: functions
./prickly-pufferfish/exec/exec/problems/generator.py:4:1: F401 'pprint.pprint' imported but unused
./prickly-pufferfish/exec/exec/problems/valid_parens/solution.py:19:46: E226 missing whitespace around arithmetic operator
./prickly-pufferfish/exec/exec/problems/valid_parens/solution.py:24:38: E226 missing whitespace around arithmetic operator
./prickly-pufferfish/exec/exec/problems/permutations/generator.py:4:1: F401 'pprint.pprint' imported but unused
./prickly-pufferfish/arena/main/models.py:1:1: F401 'django.db.models' imported but unused
./prickly-pufferfish/arena/main/admin.py:1:1: F401 'django.contrib.admin' imported but unused
```

File: prickly-pufferfish/arena/battle/consumers.py [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+            await self.del_disconnected_challenger()
+        elif self.battle.state == Battle.READY:
+            await self.del_disconnected_challenger()
+            await self.channel_layer.group_send(
+                self.battle_group_name, {
+                    'type': 'battle.unready',
+                }
+            )
+        elif self.battle.state == Battle.ACTIVE:
+            pass
```
Comment (MrHemlock) [inline prickly-pufferfish/arena/battle/consumers.py:37]: Redundant.  Doesn't need to be checked for if it's just `pass`ed.

File: prickly-pufferfish/arena/battle/views.py [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+    def name(self):
+        path = "python_questions"
+        names = os.listdir(path)
+        # removes hidden files, just in case.
+        names = [f for f in names if not f.startswith('.')]
+        name = random.choice(names)
+        return name
+
+
+my_class = RandomQuestion()
```
Comment (MrHemlock) [inline prickly-pufferfish/arena/battle/views.py:47]: Is this specifically used anywhere?  Could also use a more descriptive variable name.

File: prickly-pufferfish/exec/exec/judge/judge.py [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+from errors import ImplementationError
+
+
+class Judge:
+    """
+    Class for determining if a solution is correct.
+    It first reads function inputs from a json file, then
+    runs those inputs in both the Solution and Submission
+    class and checks for differences in output.
+    """
```
Comment (MrHemlock) [inline prickly-pufferfish/exec/exec/judge/judge.py:15]: Some good docstrings and comments in this file.  Always happy to see that!  Also great use of type hints, that's always helpful.

File: prickly-pufferfish/exec/exec/problems/generator.py [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+
+class Generator:
+    """
+    a class for generating testcases
+    """
+
+    def __init__(self):
+        pass
+
+    def generate(self, func_name: str, types: list, num=10):
```
Comment (MrHemlock) [inline prickly-pufferfish/exec/exec/problems/generator.py:19]: Is this an exact duplicate of the file at `prickly-pufferfish/exec/exec/problems/permutations/generator.py`?  If so why have both?

File: prickly-pufferfish/python_questions/add_to_zero.py [hunk 5]
```diff
@@ -0,0 +1,16 @@
+"""
+Write a function with a list of ints as a paramter. /
+Return True if any two nums sum to 0. /
+
```
Comment (MrHemlock) [inline prickly-pufferfish/python_questions/add_to_zero.py:4]: Are the questions repeated more than once?  There seems to be 3 separate places that have the questions listed or written out.

**Review by MrHemlock — COMMENTED** [review]
Thank you guys for your patience and your participating in the jam!  First thing out of the gate is that the use of Docker is a great choice!  Helps keep things consistent between system and makes it easier to get the projects up and running.

There was a fair amount of inconsistency across the project.  A handful of files had a good use of doc strings, comments and type hints.  But the majority lacked any kind of comments or information.  A fair number of the Python Questions seem to be duplicated in 3 different places (although the questions themselves look pretty fun!).  The readme file also could have used some fleshing out.

Overall it's clear that there's some hard work that was put into this project, and I think with some polish and reorganization it could really shine.  Keep these comments in mind and give it your all in the next jam as well!

**Pull request merged**